x/gamm/pool-models/balancer/pool.go

```diff
@@ -715,12 +726,15 @@ func (p *Pool) CalcJoinPoolShares(ctx sdk.Context, tokensIn sdk.Coins, swapFee s
 	// * numShares is how many shares are perfectly matched.
 	// * remainingTokensIn is how many coins we have left to join, that have not already been used.
 	// if remaining coins is empty, logic is done (we joined all tokensIn)
-	numShares, remainingTokensIn, err := cfmm_common.MaximalExactRatioJoin(p, sdk.Context{}, tokensIn)
+	numShares, tokensJoined, remainingTokensIn, err := p.CalcJoinPoolNoSwapShares(ctx, tokensIn, swapFee)
```

Comment:
I don't really understand the need for introducing a 3-value return when in any valid case we should be able to compute:
 `tokensJoined = tokensIn - remainingTokensIn`

I think that an additional return that is computed from other return values and/or arguments, only introduces unnecessary complexity.

Please let me know what you think @AlpinYukseloglu

Reply:
The original issue here was that `CalcJoinPoolShares` reimplemented a significant portion of the logic that already existed in `CalcJoinPoolNoSwapShares`.

This line happened to be a convenient place to put `CalcJoinPoolNoSwapShares`, but it's doing more than adding `tokensJoined` as a return value (e.g. it's also running the validation logic that was removed above).

My intent was to minimize rewritten code for general code simplicity and clarity. Does this change your position on this, or are you still in favor of removing the use of `CalcJoinPoolNoSwapShares` and reimplementing some of the logic here directly?

Reply:
On another look, it seems to me that we should be able to remove the `remainingTokens` in return value and have:

```
	numShares, tokensJoined, _, err = p.CalcJoinPoolNoSwapShares(ctx, tokensIn, swapFee)
	if err != nil {
		return sdk.ZeroInt(), sdk.NewCoins(), err
	}

	// safely ends the calculation if all input tokens are successfully LP'd
	if tokensJoined.IsAnyGT(tokensIn) {
		return sdk.ZeroInt(), sdk.NewCoins(), errors.New("an error has occurred, more coins joined than tokens passed in")
	} else if tokensIn.IsEqual(tokensJoined) {
		return numShares, tokensJoined, nil
	}
	remainingTokensIn := tokensIn.Sub(tokensJoined)
```

The [other caller](https://github.com/osmosis-labs/osmosis/blob/4797f8f3fc01594c3ce9fdd9f3b4b07128c74b04/x/gamm/pool-models/balancer/pool.go#L669) of `CalcJoinPoolNoSwapShares` does not currently use the third return. Since it can be computed from the second return `tokensJoined` and the argument `tokensIn`, I think we can simply keep the 2 return values.

Wdyt?

Reply:
I think removing `remainingTokens` makes sense here – let me know if my recent commit addresses your point!